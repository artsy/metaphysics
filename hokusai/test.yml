version: "2"
services:
  metaphysics:
    command: env NODE_OPTIONS=--max_old_space_size=4096 yarn ci
    extends:
      file: build.yml
      service: metaphysics
    environment:
      - MEMCACHED_URL=metaphysics-memcached:11211
      - CI_PULL_REQUEST=$CI_PULL_REQUEST
      - CIRCLE_PULL_REQUEST=$CIRCLE_PULL_REQUEST
      - CIRCLE_BUILD_NUM=$CIRCLE_BUILD_NUM
      - ENABLE_CONSIGNMENTS_STITCHING=true
    env_file: ../.env.test
    depends_on:
      - metaphysics-memcached
  metaphysics-memcached:
    image: memcached
    ports:
      - 11211:11211
