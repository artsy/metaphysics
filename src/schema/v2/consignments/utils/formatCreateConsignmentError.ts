import {
  GraphQLInt,
  GraphQLNonNull,
  GraphQLObjectType,
  GraphQLString,
} from "graphql"
import { formatGravityError } from "lib/gravityErrorHandler"
import { ResolverContext } from "types/graphql"

export const ConsignmentInquiryMutationErrorName =
  "ConsignmentInquiryMutationError"

export const ConsignmentInquiryErrorType = new GraphQLObjectType<
  any,
  ResolverContext
>({
  name: ConsignmentInquiryMutationErrorName,
  fields: () => ({
    type: {
      type: GraphQLString,
    },
    message: {
      type: new GraphQLNonNull(GraphQLString),
    },
    error: {
      type: GraphQLString,
    },
    statusCode: {
      type: GraphQLInt,
    },
  }),
})

export const formatCreateConsignmentError = (error) => {
  return formatGravityError(error)
}
