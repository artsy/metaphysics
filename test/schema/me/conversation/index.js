import { runAuthenticatedQuery } from "test/utils"

describe("Me", () => {
  describe("Conversation", () => {
    const rootValue = {
      conversationLoader: () => {
        return Promise.resolve({
          id: "420",
          initial_message: "Buncha secret stuff Message from Percy:\n\nLoved some of the works at your fair booth!",
          from_email: "collector@example.com",
          from_name: "Percy",
          _embedded: {
            last_message: {
              snippet: "Cool snippet",
              from_email_address: "other-collector@example.com",
              id: "25",
            },
          },
          items: [
            {
              item_type: "Artwork",
              item_id: "artwork-42",
              title: "Pwetty Cats",
              properties: {
                title: "Pwetty Cats",
                acquireable: true,
                artists: [
                  {
                    id: "artist-42",
                  },
                ],
              },
            },
            {
              item_type: "PartnerShow",
              item_id: "show-42",
              title: "Catty Show",
              properties: {
                is_reference: true,
                display_on_partner_profile: true,
              },
            },
          ],
        })
      },
      conversationMessagesLoader: ({ sort } = { sort: "asc" }) => {
        let messages = [
          {
            id: "240",
            raw_text: "this is a good message",
            from_email_address: "fancy_german_person@posteo.de",
            from_id: null,
            attachments: [],
            metadata: {
              lewitt_invoice_id: "420i",
            },
            from: `"Percy Z" <percy@cat.com>`,
            body: "I'm a cat",
          },
          {
            id: "241",
            raw_text: "this is a good message",
            from_email_address: "postman@posteo.de",
            from_id: null,
            attachments: [],
            metadata: {},
            from: `"Bitty Z" <Bitty@cat.com>`,
            body: "",
          },
          {
            id: "242",
            raw_text: "this is a good message",
            from_email_address: "fancy_german_person+wunderbar@posteo.de",
            from_id: "user-42",
            attachments: [],
            metadata: {},
            from: `"Matt Z" <matt@cat.com>`,
            body: null,
          },
          {
            id: "243",
            raw_text: "this is a good message",
            from_email_address: "postman+wunderlich@posteo.de",
            from_id: "user-21",
            attachments: [],
            metadata: {},
            from: "<email@email.com>",
            deliveries: [
              {
                opened_at: "2020-12-31T12:00:00+00:00",
              },
            ],
          },
        ]

        if (sort === "desc") {
          messages = messages.reverse()
        }

        return Promise.resolve({
          total_count: 4,
          message_details: messages,
        })
      },
      conversationInvoiceLoader: () =>
        Promise.resolve({
          payment_url: "https://www.adopt-cats.org/adopt-all-the-cats",
          state: "unpaid",
          symbol: "$",
          total_cents: 420000,
          lewitt_invoice_id: "420",
          id: "1",
        }),
    }

    it("returns a conversation", () => {
      const query = `
        {
          me {
            conversation(id: "420") {
              id
              initial_message
              from {
                email
              }
              last_message_id
              messages(first: 10) {
                edges {
                  node {
                    id
                    is_invoice
                    invoice {
                      __id
                      id
                      payment_url
                      state
                      total
                    }
                    is_from_user
                    from {
                      name
                      email
                    }
                    body
                    deliveries {
                      opened_at
                    }
                  }
                }
              }
            }
          }
        }
      `

      return runAuthenticatedQuery(query, rootValue).then(({ me: conversation }) => {
        expect(conversation).toMatchSnapshot()
      })
    })

    describe("concerning unread indicator", () => {
      it("returns the right unread status", () => {
        const query = `
          {
            me {
              conversation(id: "420") {
                is_last_message_to_user
                last_message_open
                last_message_delivery_id
              }
            }
          }
        `

        const message = {
          message_details: [
            {
              deliveries: [{ id: "2", email: "collector@example.com", opened_at: "2020-12-31T12:00:00+00:00" }],
            },
          ],
        }

        const customRootValue = Object.assign({}, rootValue, {
          conversationMessagesLoader: () => Promise.resolve(message),
        })

        return runAuthenticatedQuery(query, customRootValue).then(({ me: { conversation } }) => {
          expect(conversation).toMatchSnapshot()
        })
      })
    })

    describe("concerning items", () => {
      const query = `
        {
          me {
            conversation(id: "420") {
              items {
                title
                item {
                  __typename
                  ... on Artwork {
                    is_acquireable
                  }
                  ... on Show {
                    is_reference
                  }
                }
              }
            }
          }
        }
      `

      it("returns the conversation items", () => {
        return runAuthenticatedQuery(query, rootValue).then(({ me: { conversation: { items } } }) => {
          expect(items.length).toEqual(2)
          expect(items).toMatchSnapshot()
        })
      })

      it("doesnt return invalid items", () => {
        const customRootValue = Object.assign({}, rootValue, {
          // Get a copy of the conversation payload, mutate it, and return that instead.
          conversationLoader: () =>
            rootValue.conversationLoader().then(conversation => {
              const convo = conversation
              convo.items[0].properties = {}
              return convo
            }),
        })

        return runAuthenticatedQuery(query, customRootValue).then(({ me: { conversation: { items } } }) => {
          expect(items.length).toEqual(1)
          expect(items).toMatchSnapshot()
        })
      })
    })

    describe("messages", () => {
      const getQuery = (sort = "ASC") => {
        return `
          {
            me {
              conversation(id: "420") {
                messages(first: 10, sort: ${sort}) {
                  edges {
                    node {
                      id
                    }
                  }
                }
              }
            }
          }
        `
      }

      it("returns messages in ascending order", () => {
        const query = getQuery()

        return runAuthenticatedQuery(query, rootValue).then(({ me: { conversation: { messages } } }) => {
          expect(messages.edges.length).toEqual(4)
          expect(messages.edges[0].node.id).toEqual("240")
        })
      })

      it("returns messages in descending order", () => {
        const query = getQuery("DESC")

        return runAuthenticatedQuery(query, rootValue).then(({ me: { conversation: { messages } } }) => {
          expect(messages.edges.length).toEqual(4)
          expect(messages.edges[0].node.id).toEqual("243")
        })
      })
    })
  })
})
