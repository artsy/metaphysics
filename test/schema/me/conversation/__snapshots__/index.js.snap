// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Me Conversation concerning items doesnt return invalid items 1`] = `
Array [
  Object {
    "item": Object {
      "__typename": "Show",
      "is_reference": true,
    },
    "title": "Catty Show",
  },
]
`;

exports[`Me Conversation concerning items returns the conversation items 1`] = `
Array [
  Object {
    "item": Object {
      "__typename": "Artwork",
      "is_acquireable": true,
    },
    "title": "Pwetty Cats",
  },
  Object {
    "item": Object {
      "__typename": "Show",
      "is_reference": true,
    },
    "title": "Catty Show",
  },
]
`;

exports[`Me Conversation concerning unread indicator returns the right unread status 1`] = `
Object {
  "is_last_message_to_user": true,
  "last_message_delivery_id": "2",
  "last_message_open": "2020-12-31T12:00:00+00:00",
}
`;

exports[`Me Conversation returns a conversation 1`] = `
Object {
  "conversation": Object {
    "from": Object {
      "email": "collector@example.com",
    },
    "id": "420",
    "initial_message": "Loved some of the works at your fair booth!",
    "last_message_id": "25",
    "messages": Object {
      "edges": Array [
        Object {
          "node": Object {
            "body": "I'm a cat",
            "deliveries": null,
            "from": Object {
              "email": "fancy_german_person@posteo.de",
              "name": "Percy Z",
            },
            "id": "240",
            "invoice": Object {
              "__id": "SW52b2ljZTox",
              "id": "1",
              "payment_url": "https://www.adopt-cats.org/adopt-all-the-cats",
              "state": "UNPAID",
              "total": "$4,200",
            },
            "is_from_user": false,
            "is_invoice": true,
          },
        },
        Object {
          "node": Object {
            "body": "",
            "deliveries": null,
            "from": Object {
              "email": "postman@posteo.de",
              "name": "Bitty Z",
            },
            "id": "241",
            "invoice": null,
            "is_from_user": false,
            "is_invoice": false,
          },
        },
        Object {
          "node": Object {
            "body": null,
            "deliveries": null,
            "from": Object {
              "email": "fancy_german_person+wunderbar@posteo.de",
              "name": "Matt Z",
            },
            "id": "242",
            "invoice": null,
            "is_from_user": true,
            "is_invoice": false,
          },
        },
        Object {
          "node": Object {
            "body": null,
            "deliveries": Array [
              Object {
                "opened_at": "2020-12-31T12:00:00+00:00",
              },
            ],
            "from": Object {
              "email": "postman+wunderlich@posteo.de",
              "name": null,
            },
            "id": "243",
            "invoice": null,
            "is_from_user": false,
            "is_invoice": false,
          },
        },
      ],
    },
  },
}
`;
